### PR TITLE
[WIP] ARROW-2910: [Packaging] Build from official apache archive

### PR DESCRIPTION
**Pull request by kszucs**
Add an optional `--arrow-dist` argument to `submit` command, which will be downloaded and checksum verified on each builds.
- [ ] all of the wheels are working
- [x] all of the conda packages are working
- [ ] all of the linux packages are working

Waiting for build 286 and 287 (from git, from archive) https://github.com/kszucs/crossbow/branches

### CODE REVIEW

**Comment by kszucs** [comment]
Awkward, the sha512 checksum verification fails on windows: https://ci.appveyor.com/project/kszucs/crossbow/build/1.0.364#L9
It matches on both osx and linux.

**Comment by kszucs** [comment]
@cpcloud @wesm I'm afraid this is going to take more time, there are a lot of hardcoded behaviour and paths. Testing is painful, consumes a lot of time. 

Ideally We either build from git (this is functional on arrow master):
1. Clone Arrow
2. Build a source archive (currently this is done only for [linux packages](https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/Rakefile#L72) - others use the git repo directly)
3. Build the particular binaries

Or build from apache source archive (implemented by this PR):
1. Download source archive (produced by `02-source.sh`)
2. Build the particular binaries (`wheel-linux-*` and `linux-*` builds are not yet working with this setup)

The linux Rakefile already handles multiple sources (git + build the source archive, rc archive, release archive), which We should turn off. Actually We don't have arrow's repository around when it builds from release archive, but the Rakefile [expects](https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/Rakefile#L24) the git repo too. 

@kou Could You refactor the rake tasks to support both methods and receive all of the required parameters via arguments and/or env variables (in the first case We'd have a directory containing a git repository, in the second case We'd have the [extracted source archive](https://github.com/apache/arrow/pull/2323/files#diff-66ceaeef0ef768c9dcd99e8da32ce261R43)). 

Also shouldn't We update the [changelog](https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/package-task.rb#L298) etc. in `02-source.sh`?

**Comment by wesm** [comment]
@cpcloud do you have some time to help with this? I think we're going to need more hands. I can spend some time on it the next few days also. 

The changelog is updated in https://github.com/apache/arrow/blob/master/dev/release/00-prepare.sh#L30, is that causing an issue?

**Comment by wesm** [comment]
We should probably also try to distill from this experience how we might have applied a testing-based / TDD development strategy to improve dev iteration times. One of the big problems here is that we are "testing in production" which is very time consuming

**Comment by kou** [comment]
@kszucs I've fixed linux-packages build! If we have `dev/tasks/linux-packages/apache-arrow-#{version}.tar.gz`, `rake dist` does nothing.

**Comment by kou** [comment]
> The changelog is updated in https://github.com/apache/arrow/blob/master/dev/release/00-prepare.sh#L30, is that causing an issue?

It makes sense.

**Comment by kou** [comment]
@kszucs I've pushed one more commit https://github.com/apache/arrow/pull/2323/commits/339624e58d482854e0daf35b6e1b977ea57d7610 because the current implementation doesn't support deb/rpm package names for RC.

I noticed that GitHub release doesn't support "~" for filename. GitHub pages renames "~" with ".". We use "~" for deb packages. Maybe, it doesn't cause a problem...

**Comment by kou** [comment]
I'll rebase on master because my commit causes a conflict.

**Comment by kou** [comment]
If we merge #2331, we don't need to run `rake version:update` for release build because it's done in `00-prepare.sh`. How about the following changes?

```diff
diff --git a/dev/tasks/crossbow.py b/dev/tasks/crossbow.py
index 2c57a9ca..4ccc3db1 100755
--- a/dev/tasks/crossbow.py
+++ b/dev/tasks/crossbow.py
@@ -304,6 +304,10 @@ class Target:
         else:
             return '{}.0.{}'.format(base_version, rc)
 
+    @property
+    def is_release(self):
+        return re.match(r'\d+\.\d+\.\d+\Z', self.version) is not None
+
     def _split_version(self):
         match = re.match(r'(.+)-(rc\d+)\Z', self.version)
         if match is None:
diff --git a/dev/tasks/linux-packages/travis.linux.yml b/dev/tasks/linux-packages/travis.linux.yml
index ec89cdbc..1f9cced6 100644
--- a/dev/tasks/linux-packages/travis.linux.yml
+++ b/dev/tasks/linux-packages/travis.linux.yml
@@ -69,7 +69,9 @@ install:
 
 script:
   - pushd arrow/dev/tasks/linux-packages
+{%- if not arrow.is_release %}
   - rake version:update
+{%- endif %}
 {%- if arrow.source == 'git' %}
   - rake dist
 {%- endif %}
```

**Comment by kszucs** [comment]
@kou With crossbow We pass version explicitly, so actually We don't need to run `version:update` at all.

**Comment by kszucs** [comment]
FYI Testing from git:
```bash
python crossbow.py submit -g linux
```

Testing via source archive:
```bash
# create source archive on github releases
cd arrow_git_repo
git tag -d apache-arrow-0.10.0 
git tag apache-arrow-0.10.0
git push --tags -f

# submit build from tarball without checksum verification
python crossbow.py submit \
    --no-verify \
    --arrow-dist https://github.com/kszucs/arrow/archive/apache-arrow-0.10.0.tar.gz  \
    -g linux
```

**Comment by kszucs** [comment]
My time and internet access will be limited during the weekend, but I'm trying to be available.

I've just submitted two [jobs](https://github.com/kszucs/crossbow/branches/yours) with all of the tasks:
- git: https://github.com/kszucs/crossbow/tree/build-280
- archive: https://github.com/kszucs/crossbow/tree/build-279

@kou 
centos-6 from git works: https://travis-ci.org/kszucs/crossbow/builds/408876960
centos-6 from source fails: https://travis-ci.org/kszucs/crossbow/builds/408877257

This might be because I'm using github release archives, not building one with prepare.sh and source.sh scripts.

**Comment by wesm** [comment]
Looks like there's some issue with the directory path

```
+ cd apache-arrow-0.10.0
/var/tmp/rpm-tmp.kjpWL5: line 38: cd: apache-arrow-0.10.0: No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.kjpWL5 (%prep)
```

**Comment by kou** [comment]
Yes. It's a directory path problem.
https://github.com/kszucs/arrow/archive/apache-arrow-0.10.0.tar.gz uses `arrow-apache-arrow-0.10.0` instead of `apache-arrow-0.10.0` (no `arrow-` prefix).

**Comment by kou** [comment]
I've pushed a commit https://github.com/apache/arrow/pull/2323/commits/039b88c6310f3677c0fe12eff5c086cfcb41337d that stops to require "make dist"-ed source archive. It means that we can use `.tar.gz` created by `git archive` if `.tar.gz` uses `apache-arrow-XXX` for the top directory.

**Comment by kou** [comment]
I'll rebase on master to resolve conflict with https://github.com/apache/arrow/commit/0707b16e54b1a5471b49282c557ddb06f8dacc35 .

**Comment by kou** [comment]
Done.

File: dev/tasks/crossbow.py [hunk 1]
```diff
@@ -301,6 +343,56 @@ def from_repo(cls, repo):
...
+        return 'dist'
+
+    @property
+    def signature_url(self):
+        return self.archive_url + '.asc'
+
+    @property
+    def checksum_url(self):
+        """SHA512 checksum"""
+        return self.archive_url + '.sha512'
```
Comment (cpcloud) [inline dev/tasks/crossbow.py]: There are going to be multiple checksums here: sha1, sha256, and sha512 so we should handle that.

File: dev/tasks/crossbow.py [hunk 2]
```diff
@@ -443,24 +536,46 @@ def load_tasks_from_config(config_path, task_names, group_names):
...
+            'deb_version': target.deb_version,
+            'rpm_version': target.rpm_version,
+        }
         artifacts = task.pop('artifacts', None) or []  # because of yaml
-        artifacts = [fn.format(version=target.version) for fn in artifacts]
+        artifacts = [fn.format(**variables) for fn in artifacts]
         tasks[name] = Task(**task, artifacts=artifacts)
 
     # create job instance, doesn't mutate git data yet
+    if len(tasks) == 0:
```
Comment (cpcloud) [inline dev/tasks/crossbow.py]: Can you make this `not tasks` instead?

File: dev/tasks/python-wheels/appveyor.yml [hunk 3]
```diff
@@ -25,17 +27,31 @@ environment:
   MSVC_DEFAULT_OPTIONS: ON
   ARROW_SRC: C:\apache-arrow
   PYARROW_VERSION: {{ arrow.version }}
-  PYARROW_REF: {{ arrow.head }}
-  PARQUET_CPP_REF: master
+  PARQUET_CPP_REF: apache-parquet-cpp-1.4.0
```
Comment (cpcloud) [inline dev/tasks/python-wheels/appveyor.yml]: This needs to be remain master for now because of ARROW-2929.

File: dev/tasks/crossbow.py (hunk 1 above)
Comment (cpcloud) [reply dev/tasks/crossbow.py]: Can you change this to `.sha256`? As of ARROW-2934 we will only be generating `.sha256` and not `.sha512`.

Comment (kszucs) [reply dev/tasks/crossbow.py]: sure

**Comment by kszucs** [comment]
@cpcloud We should create a crossbow script to create source archives, either submitting them to github releases (for testing) or apache svn. As @kou has mentioned github source archives have different names, which causes linux builds to fail. Also `-rcN` archive have special treatment, which again causes problems.

**Comment by wesm** [comment]
Closing as stale

**Pull request closed without merging**